extensions/amp-document-recommendations/0.1/amp-document-recommendations.js

```diff
@@ -39,20 +40,24 @@ const MAX_ARTICLES = 2;
  */
 const SEPARATOR_RECOS = 3;
 
-/** @private {AmpDocumentRecommendations} */
-let activeInstance_ = null;
+const SERVICE_ID = 'document-recommendations';
 
 export class AmpDocumentRecommendations extends AMP.BaseElement {
 
   /** @param {!AmpElement} element */
   constructor(element) {
     super(element);
 
-    // TODO(emarchiori): Consider using a service instead of singleton.
-    if (activeInstance_) {
+    /** @private {boolean} */
+    this.isActive_ = false;
+
+    const service = getExistingServiceOrNull(this.win, SERVICE_ID);
+    if (service !== null) {
```

Comment:
Early return from a constructor is an unusual pattern. I don't think you need to guard against multiple instances if you register it as a service, since the service manager will complain if it's registered twice. Do any of our other services have a similar `isActive_` and shortcut return pattern?

Reply:
Yeah, it is weird. Removed the early return.

We need a separate check for multiple instances as we only allow a single instance of this component to register per window, so we need to not run the buildCallback on any other instances. We don't want to end up with an infinite scrolling page within an infinite scrolling page, as they'd end up fighting.

Reply:
What purpose does the `<amp-document-recommendations>` tag serve? Configuration and as a marker for where to insert the new document? I think the best way forward for this component to avoid reimplementing the service singletons would be to install a service more like how `amp-bind` does it. It has an `amp-state` tag for configuring the internal state, and installs the `bind-impl` service singleton for global logic. 

https://github.com/ampproject/amphtml/blob/master/extensions/amp-bind/0.1/amp-bind.js for registration
https://github.com/ampproject/amphtml/blob/master/extensions/amp-bind/0.1/bind-impl.js for singleton
https://github.com/ampproject/amphtml/blob/master/extensions/amp-bind/0.1/amp-state.js for element config

Reply:
Added another commit which creates a new class to use as the service singleton. lmk what you think.

We need to use the first `<amp-document-recommendations>` tag element to append new documents into later, which means the service is doing DOM manipulation on the component's element. Are there other services which do a similar thing?